updatedAt 필드를 이용한 최신 업데이트 순 정렬 구현

ahyeonkong opened the pull request:
## #️⃣ Issue Number

#80 

## 📝 작업 설명

updatedAt 필드를 이용해서 가장 최근에 변경된 상품이 가장 위에 올라오도록 구현했습니다.

## 📸스크린샷 (선택)   

## 💬 리뷰 요구사항

원래 bumpAt, addedDate 필드를 이용해서 정렬을 했었는데, 이때 새 상품 생성을 하면 끌어올리기 밑으로 상품이 생성되는 문제가 있었습니다.
updatedAt 필드를 이용한다면 새 상품 생성 시에도 끌어올리기 위로 상품이 생성됩니다!

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

ahyeonkong committed:
feat: updatedAt 필드를 이용한 최신 업데이트 순 정렬 구현

JinTaekLim commented on src/main/java/com/market/saessag/domain/product/service/ProductService.java:
```diff
@@ -56,6 +54,7 @@ public ProductResponse createProduct(ProductRequest productRequest, HttpServletR
         User user = userRepository.findById(userSession.getId())
                 .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다."));
 
+        LocalDateTime now = LocalDateTime.now();
```
위와같은 방식으로 작성하게 되면, bumpAt/updatedAt 뿐 아닌 다른 데이터가 실수로 변경될 수도 있을 거라 생각하는데 다른 방법은 없을까요?

ahyeonkong replied:
위처럼 수동으로 데이터를 변경한다면, updatedAt이 아닌 addedDate가 변경되는 실수가 발생할 수도 있겠네요!

Spring Data JPA에서 제공하는 JPA Auditing을 사용하면 자동으로 데이터를 업데이트 할 수 있습니다.

1. **AuditingEntityListener**를 통해 **@CreatedDate, @LastModifiedDate**를 이용하는 방법
2. 별도의 **Auditable 인터페이스**를 사용하는 방법

이런 방법들이 있을 것 같네요!

ahyeonkong replied:
추가로 빌더와 정적 팩토리 메서드 패턴을 동시에 활용하면 해당 문제를 해결할 수 있을 것 같네요!

JinTaekLim commented on src/main/java/com/market/saessag/domain/product/service/ProductService.java:
```diff
@@ -128,12 +129,13 @@ public boolean deleteProduct(Long productId, HttpServletRequest request) {
-                    Sort.Order.desc("bumpAt"),
-                    Sort.Order.desc("addedDate")
+                    Sort.Order.desc("updatedAt"),  // 최신 업데이트 순
+                    Sort.Order.desc("bumpAt")      // 끌어올리기 순
```
updatedAt 을 기준으로 정렬하게 되면, 상품 생성시에도 updatedAt을 업데이트해야할 것같습니다!
추가적으로 정렬 기준에 bumpAt을 포함한 이유가 있을까요?

ahyeonkong replied:
넵 현재 상품 생성 시 updatedAt을 업데이트 하고 있습니다!
기존 코드에 bumpAt이 작성되어 있어서 그대로 두었는데, 모든 업데이트 내역은 updatedAt 필드 하나에서만 관리하는 게 좋을 것 같아요!

ahyeonkong committed:
Merge branch 'develop' of https://github.com/saessagMarket/back into feat/#80-product-sort-desc

ahyeonkong committed:
refactor: ProductRequest에 Lombok 어노테이션 추가

- @Getter, @Builder, @NoArgsConstructor, @AllArgsConstructor

ahyeonkong committed:
refactor: 상품 엔티티 리팩토링

- bumpAt 필드 제거(updatedAt 필드로 통합)
- 상품 생성을 위한 createProduct 정적 팩토리 메서드 추가
- 상품 수정을 위한 updateProduct 정적 팩토리 메서드 추가
- 불필요한 updateBumpAt, updateUpdatedAt 메서드 제거

ahyeonkong committed:
refactor: 상품 서비스 로직 개선

- 세션 검증 및 사용자 조회 공통 로직 getUserFromSession 메서드로 분리
- 상품 조회 및 소유자 검증 공통 로직 getProductAndValidateOwner 메서드로 분리
- 상품 정렬 기준을 bumpAt에서 updatedAt으로 통합

ahyeonkong committed:
refactor: API 응답 타입 명확화 및 세션 검증 책임 분리

- ApiResponse<?> 제네릭 타입을 ApiResponse<ProductResponse>로 구체화
- 세션 검증 및 사용자 조회는 service 단에서 수행

ahyeonkong committed:
test: 상품 끌어올리기 테스트 코드 개선

- 세션 인증 관련 모킹 추가

ahyeonkong committed:
refactor: 상품 끌어올리기 로직 개선

- Product 엔티티에 bump 메서드 추가하여 updatedAt 필드만 갱신하도록 변경
- 불필요한 데이터 변경 위험 제거
